pkg/pillar/cmd/volumemgr/volumemgr.go

```diff
@@ -87,6 +92,9 @@ type volumemgrContext struct {
 	casClient cas.CAS
 
 	volumeConfigCreateDeferredMap map[string]*types.VolumeConfig
+
+	zfsVolMap      map[string]string // /dev/zd* -> pool/name
+	zfsVolMapMutex sync.Mutex
```

Comment:
The tricky bit in keeping these types of persistent data structures in sync with the state of the world is that in our architecture the individual microservices may conceptually go down and come up. I'd really like @eriknordmark to take a look at this approach first.

Reply:
@rvs The persistent and restart aspects might be OK here since processZvols adds all of the info to the map (but it doesn't seem to delete entries which have disappeared). But the retry+sleep loop below is a bit odd.

I think it would be easier to have some periodic task or goroutine which would collect the mapping from zfs and publish it using pubsub, and then have volumemgr subscribe to that publication. That way there is no need for sleep but the handlecreate/modify functions for that subscription can do the work when the information is available.

FWIW we do something similar when interacting with netlink in the kernel, but there we can create a listener socket which receives notifications (which avoids periodically checking if anything changed). I don't know if zfs has such a notification mechanism.

Reply:
Good points, @eriknordmark -- ZFS communicates via udev events so conceptually the same approach could be possible. For now, though, it seems that @giggsoff and I agreed on keeping it outside of our Go code and rely on mdev manipulating the symlinks in /dev

Reply:
I put mdev magic in my latest commits